enterprise/backend/test/metabase_enterprise/serialization/cmd_test.clj

```diff
@@ -77,4 +87,117 @@
                                               (assoc user :password "test-password"))]
           (cmd/load dump-dir "--mode"     :update
                              "--on-error" :abort)
-          (t/is (true? @user-pre-insert-called?)))))))
+          (is (true? @user-pre-insert-called?)))))))
+
+(defn- create! [model & {:as properties}]
+  (db/insert! model (merge (t.test/with-temp-defaults model) properties)))
+
+(defn- do-with-in-memory-h2-db [db-name-prefix f]
+  (let [db-name           (str db-name-prefix (mt/random-name))
+        connection-string (format "jdbc:h2:mem:%s" db-name)
+        data-source       (mdb.data-source/raw-connection-string->DataSource connection-string)]
+    ;; DB should stay open as long as `conn` is held open.
+    (with-open [_conn (.getConnection data-source)]
+      (letfn [(do-with-app-db [thunk]
+                (binding [mdb.conn/*application-db* (mdb.conn/application-db :h2 data-source)]
+                  (println (u/format-color :yellow "USING %s" (pr-str connection-string))) ; NOCOMMIT
+                  (testing (format "\nApp DB = %s" (pr-str connection-string))
+                    (thunk))))]
+        (do-with-app-db
+         (fn []
+           (mdb/setup-db!)))
+        (f do-with-app-db)))))
+
+(defn- do-with-source-and-dest-dbs [f]
+  (do-with-in-memory-h2-db
+   "source-"
+   (fn [do-with-source-db]
+     (do-with-in-memory-h2-db
+      "dest-"
+      (fn [do-with-dest-db]
+        (f do-with-source-db do-with-dest-db))))))
+
+(defmacro ^:private with-source-and-dest-dbs {:style/indent 0} [& body]
+  `(do-with-source-and-dest-dbs
+    (fn [~'&do-with-source-db ~'&do-with-dest-db]
+      ~@body)))
+
+(defmacro ^:private with-source-db {:style/indent 0} [& body]
+  `(~'&do-with-source-db (fn [] ~@body)))
+
+(defmacro ^:private with-dest-db {:style/indent 0} [& body]
+  `(~'&do-with-dest-db (fn [] ~@body)))
```

Comment:
i've been struggling to find out what this was and its an anaphor introduced by the wrapping macro `with-source-and-dest-dbs`?

Reply:
Yeah this isn't super clear. I added a bunch of docstrings and comments explaining what's going on.

Reply:
It took me a second but i really like the style of it. This is a really nice solution.